### PR TITLE
[fix](be) Preserve floating point values in CASE branch selection

### PR DESCRIPTION
**Pull request by HappenLee**
### What problem does this PR solve?

Issue Number: N/A

Related PR: N/A

Problem Summary: A non-nullable FLOAT/DOUBLE CASE can return NaN for a finite selected branch when an unselected branch contains Infinity or NaN. For example, with ordinary CASE evaluation, selecting `1`, an overflowing multiplication, and `2` over three rows returns `NaN, Infinity, NaN` instead of `1, Infinity, 2`. The same result assembly also loses the sign of selected negative zero.

The result assembly multiplies each branch value by a zero/one mask and adds it to the result. IEEE-754 arithmetic makes `0 * Infinity` and `0 * NaN` equal NaN. Use conditional stores to copy the selected floating point value without arithmetic. DATE, DATETIME, DATEV2, DATETIMEV2, TIMESTAMP_NS and TIMESTAMPTZ share this conditional-store loop, replacing their previous ternary assignments. This form allows AVX2 masked loads/stores, while a ternary assignment can become a conditional pointer load that inhibits vectorization.

Retain the existing default-value initialization and the arithmetic accumulator for the other numeric types. Removing the redundant initialization for the direct-store types is deferred.

Add bitwise unit tests, SQL regression tests and a benchmark that directly calls the production result assembly function. The typed unit tests cover all eight types with both uint8_t and uint16_t branch indices.

### Release note

Fix incorrect FLOAT/DOUBLE CASE results caused by unselected non-finite branch values, and preserve selected negative zero.

### Check List (For Author)

- Test
    - [x] Regression test
    - [x] Unit Test
    - [x] Manual test (details below)
    - [ ] No need to test or manual test.

  Validation of the date/time follow-up:
  - All 48 `VCaseSelectionTest*` tests pass via `./run-be-ut.sh -j 48 --run --filter='VCaseSelectionTest*'` with ASAN. Coverage includes all eight types, both index widths, constants, boundary values, vector tails and 255/257 branches.
  - clang-format 16, header hygiene, `git diff --check` and clang-tidy pass. For clang-tidy, the header was explicitly mapped to the generated selection-test compiler flags to avoid failed automatic header inference.
  - `./build.sh --be -j 48` with ASAN compiled and linked the BE successfully. Full build/packaging did not pass: the later `kuromoji_build_dict` process reported an ASAN double-free during OpenBLAS/OpenMP initialization (`dlsym` / `_dlerror_run`), before entering main. The separate standard ASAN unit-test build and execution passed.
  - Default-value initialization and the remaining arithmetic accumulator are unchanged.

  Earlier validation of the floating point fix (before the date/time follow-up):
  - All 12 original floating point unit tests fail on the original implementation and pass with ASAN after the fix. These cases are retained in the expanded typed suite. Coverage includes non-finite values, signed zero and subnormal values.
  - `test_case_float_nonfinite` and `test_short_circuit_evaluation` pass. The original implementation fails the new SQL regression. Golden output was generated and independently verified through short-circuit evaluation using the regression runner. These SQL suites were not rerun for the date/time follow-up.
  - ASAN BE build, clang-format 16, header hygiene and clang-tidy checks pass for the earlier revision.
  - RELEASE benchmarks built with Clang 21.1.8 and `-O3 -msse4.2 -mavx2` on Xeon Platinum 8457C cover 42 scenarios. With identical input, a fixed CPU, five repetitions per run and before/after/after/before ordering, median CPU time decreases by approximately 5%–39% (18.5% geometric mean reduction). This measures floating point result assembly on this AVX2 machine, including result allocation; these figures do not measure the date/time follow-up.
  - Disassembly of all four floating point/index-width specializations in the linked binaries confirms `vmaskmovps/pd`, replacing floating point multiply/add instructions without fast-math.

- Behavior changed:
    - [ ] No.
    - [x] Yes. Return the selected floating point value without contamination from other branches or loss of negative zero. Date/time selection preserves the existing results.

- Does this need documentation?
    - [x] No.
    - [ ] Yes.

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by hello-stephen** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://doris.apache.org/community/how-to-contribute/review-and-ci).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by HappenLee** [comment]
run buildall

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 16989 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit 0ff005b8d0bbaed15e435ebb92ce9b39502ed362, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17651	2980	3001	2980
q2	2061	253	216	216
q3	10275	882	522	522
q4	4670	267	206	206
q5	7657	583	389	389
q6	139	121	98	98
q7	529	502	402	402
q8	9239	907	890	890
q9	3459	2439	2433	2433
q10	6545	868	740	740
q11	401	202	183	183
q12	626	266	204	204
q13	18139	1598	1192	1192
q14	163	158	143	143
q15	q16	462	406	385	385
q17	1396	918	806	806
q18	3194	2342	2313	2313
q19	1284	929	815	815
q20	374	290	201	201
q21	5607	1640	1897	1640
q22	333	272	231	231
Total cold run time: 94204 ms
Total hot run time: 16989 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	3371	3283	3296	3283
q2	519	421	393	393
q3	2266	2360	2281	2281
q4	1255	1244	927	927
q5	2253	2177	2159	2159
q6	177	122	90	90
q7	1034	962	895	895
q8	1606	1410	1410	1410
q9	3261	3228	3214	3214
q10	1974	1925	1719	1719
q11	361	275	263	263
q12	465	446	352	352
q13	1521	1562	1185	1185
q14	186	173	161	161
q15	q16	406	406	364	364
q17	3678	3376	3249	3249
q18	5086	4579	5161	4579
q19	935	856	864	856
q20	1065	1012	877	877
q21	3937	3227	3251	3227
q22	409	352	330	330
Total cold run time: 35765 ms
Total hot run time: 31814 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 83147 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit 0ff005b8d0bbaed15e435ebb92ce9b39502ed362, data reload: false

query5	4246	409	324	324
query6	385	134	120	120
query7	4976	400	229	229
query8	297	127	133	127
query9	8693	2880	2841	2841
query10	390	228	208	208
query11	5399	1068	932	932
query12	113	71	80	71
query13	1185	460	329	329
query14	6084	2237	2119	2119
query14_1	2029	2003	1998	1998
query15	174	123	114	114
query16	912	365	346	346
query17	808	422	349	349
query18	2324	311	228	228
query19	160	136	100	100
query20	75	68	71	68
query21	199	97	83	83
query22	5613	5471	5628	5471
query23	6916	6437	6193	6193
query23_1	6296	6185	6199	6185
query24	7284	1091	762	762
query24_1	787	769	761	761
query25	446	274	221	221
query26	1233	224	128	128
query27	2785	374	255	255
query28	4726	1485	1499	1485
query29	915	417	323	323
query30	241	152	127	127
query31	819	399	332	332
query32	125	74	72	72
query33	443	202	174	174
query34	992	815	495	495
query35	401	412	355	355
query36	577	571	555	555
query37	118	75	67	67
query38	1031	855	852	852
query39	479	464	466	464
query39_1	485	465	465	465
query40	195	86	73	73
query41	53	54	50	50
query42	72	74	67	67
query43	237	243	214	214
query44	983	519	533	519
query45	112	107	102	102
query46	757	803	517	517
query47	757	771	704	704
query48	301	298	227	227
query49	537	228	180	180
query50	739	256	200	200
query51	8036	8018	8052	8018
query52	68	72	59	59
query53	189	209	197	197
query54	230	168	166	166
query55	73	63	56	56
query56	210	189	171	171
query57	677	711	676	676
query58	206	177	170	170
query59	1247	1274	1150	1150
query60	256	204	199	199
query61	162	128	139	128
query62	371	197	180	180
query63	172	146	141	141
query64	2829	782	665	665
query65	1660	1660	1649	1649
query66	1763	253	207	207
query67	10049	10035	10008	10008
query68	2999	1126	660	660
query69	343	223	200	200
query70	670	637	618	618
query71	246	180	163	163
query72	2300	1732	1558	1558
query73	611	587	353	353
query74	1964	1233	1147	1147
query75	1185	1122	975	975
query76	2351	705	523	523
query77	254	267	212	212
query78	4070	3801	3410	3410
query79	2699	835	612	612
query80	1575	321	270	270
query81	517	158	136	136
query82	618	123	97	97
query83	282	206	192	192
query84	288	108	85	85
query85	777	327	276	276
query86	479	177	168	168
query87	1030	989	893	893
query88	2935	2109	2083	2083
query89	295	193	174	174
query90	1986	123	123	123
query91	131	115	98	98
query92	93	66	70	66
query93	1740	1078	644	644
query94	629	240	225	225
query95	538	330	226	226
query96	771	553	272	272
query97	1086	1056	1036	1036
query98	168	139	132	132
query99	415	349	307	307
Total cold run time: 179550 ms
Total hot run time: 83147 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 14.76 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit 0ff005b8d0bbaed15e435ebb92ce9b39502ed362, data reload: false

query1	0.01	0.00	0.01
query2	0.08	0.04	0.04
query3	0.25	0.11	0.11
query4	1.60	0.09	0.10
query5	0.17	0.15	0.15
query6	1.28	0.71	0.68
query7	0.03	0.00	0.01
query8	0.05	0.03	0.03
query9	0.29	0.21	0.21
query10	0.38	0.36	0.34
query11	0.16	0.11	0.12
query12	0.14	0.12	0.11
query13	0.33	0.31	0.30
query14	0.45	0.46	0.46
query15	0.37	0.35	0.36
query16	0.20	0.22	0.22
query17	0.70	0.70	0.68
query18	0.18	0.17	0.17
query19	1.15	1.11	1.14
query20	0.02	0.01	0.01
query21	15.44	0.15	0.13
query22	5.06	0.04	0.04
query23	16.17	0.26	0.11
query24	3.01	0.33	0.26
query25	0.12	0.04	0.04
query26	0.84	0.16	0.12
query27	0.04	0.03	0.04
query28	3.68	0.56	0.28
query29	12.47	3.22	2.56
query30	0.26	0.12	0.12
query31	2.76	0.38	0.18
query32	3.50	0.32	0.23
query33	1.47	1.40	1.52
query34	15.34	2.18	1.81
query35	1.77	1.81	1.75
query36	0.46	0.29	0.30
query37	0.06	0.04	0.04
query38	0.05	0.03	0.03
query39	0.03	0.03	0.03
query40	0.12	0.07	0.08
query41	0.07	0.02	0.02
query42	0.03	0.03	0.02
query43	0.03	0.03	0.03
Total cold run time: 90.62 s
Total hot run time: 14.76 s
```
</details>

File: be/src/exprs/vcase_expr.h [hunk 1]
```diff
@@ -249,12 +249,22 @@ class VCaseExpr final : public VExpr {
...
-                          std::is_same_v<ColumnType, ColumnDateTimeV2> ||
-                          std::is_same_v<ColumnType, ColumnTimeStampNs> ||
-                          std::is_same_v<ColumnType, ColumnTimeStampTz>) {
+            if constexpr (std::is_same_v<ColumnType, ColumnFloat32> ||
+                          std::is_same_v<ColumnType, ColumnFloat64>) {
+                // Arithmetic masking propagates unselected NaN/Infinity and loses signed zero.
+                // Conditional stores also let the compiler vectorize without loading from a
+                // selected source/destination pointer, as a ternary assignment can do.
+                for (size_t row_idx = 0; row_idx < rows_count; row_idx++) {
+                    if (then_idx[row_idx] == i) {
```
[the remark this answers is not in the document]
Comment (HappenLee) [reply be/src/exprs/vcase_expr.h:264]: Updated in [d82e48a99dc](https://github.com/apache/doris/pull/67896/commits/d82e48a99dc60c4e692e53ed141210bcbf362df5): DATE, DATETIME, DATEV2, DATETIMEV2, TIMESTAMP_NS and TIMESTAMPTZ now share the FLOAT/DOUBLE conditional-store loop. All 48 ASAN selection tests pass, covering all eight types with both index widths, constants, boundary values, vector tails and 255/257 branches.

The default-value initialization is intentionally retained in this PR. It is redundant for these direct-store specializations, but removing it is deferred; the remaining accumulator paths still require initialization. This update does not remove the zero-fill requested by this comment. Resolving the thread with that scope decision.

**Comment by HappenLee** [comment]
run buildall

**Comment by hello-stephen** [comment]
# BE UT Coverage Report
Increment line coverage `100.00% (4/4)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/d82e48a99dc60c4e692e53ed141210bcbf362df5_d82e48a99dc60c4e692e53ed141210bcbf362df5/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/d82e48a99dc60c4e692e53ed141210bcbf362df5_d82e48a99dc60c4e692e53ed141210bcbf362df5/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 63.30% (29655/46847) |
| Line Coverage     | 48.23% (310041/642772) |
| Region Coverage   | 43.76% (249603/570403) |
| Branch Coverage   | 45.36% (116236/256266) |

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-H: <b>Total hot run time: 16782 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpch-tools
Tpch sf100 test result on commit d82e48a99dc60c4e692e53ed141210bcbf362df5, data reload: false

------ Round 1 ----------------------------------
============================================
q1	17595	3043	3032	3032
q2	2079	268	236	236
q3	10223	887	510	510
q4	4672	248	199	199
q5	7678	587	382	382
q6	138	115	97	97
q7	535	494	393	393
q8	9245	889	916	889
q9	3446	2385	2370	2370
q10	6539	860	690	690
q11	391	196	184	184
q12	608	261	196	196
q13	18124	1514	1162	1162
q14	160	148	138	138
q15	q16	438	398	374	374
q17	1346	838	787	787
q18	3100	2271	2261	2261
q19	1274	843	709	709
q20	364	281	194	194
q21	5606	1750	1890	1750
q22	319	271	229	229
Total cold run time: 93880 ms
Total hot run time: 16782 ms

----- Round 2, with runtime_filter_mode=off -----
============================================
q1	3408	3336	3311	3311
q2	497	385	368	368
q3	2252	2305	2188	2188
q4	1189	1156	893	893
q5	2174	2099	2081	2081
q6	175	122	86	86
q7	1033	989	832	832
q8	1614	1386	1388	1386
q9	3144	3100	3080	3080
q10	1864	1809	1626	1626
q11	355	267	253	253
q12	457	428	347	347
q13	1472	1548	1163	1163
q14	165	186	164	164
q15	q16	388	405	354	354
q17	3589	3287	3211	3211
q18	4795	4414	4730	4414
q19	878	826	928	826
q20	1025	944	831	831
q21	3784	3126	3295	3126
q22	383	341	315	315
Total cold run time: 34641 ms
Total hot run time: 30855 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>TPC-DS: <b>Total hot run time: 81783 ms</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/tpcds-tools
TPC-DS sf100 test result on commit d82e48a99dc60c4e692e53ed141210bcbf362df5, data reload: false

query5	4237	429	325	325
query6	414	132	128	128
query7	4912	421	230	230
query8	291	128	120	120
query9	8697	2885	2885	2885
query10	407	231	187	187
query11	5392	1021	899	899
query12	121	71	71	71
query13	1219	450	319	319
query14	6028	2176	2073	2073
query14_1	1944	1942	1930	1930
query15	178	120	108	108
query16	911	373	343	343
query17	772	434	341	341
query18	2317	296	230	230
query19	159	130	107	107
query20	68	67	71	67
query21	195	101	85	85
query22	5550	5464	5356	5356
query23	6721	6178	5957	5957
query23_1	6049	6179	5985	5985
query24	7338	1087	778	778
query24_1	778	787	814	787
query25	419	297	256	256
query26	1238	231	133	133
query27	2781	411	233	233
query28	4706	1502	1531	1502
query29	923	437	355	355
query30	245	157	133	133
query31	811	403	334	334
query32	125	77	71	71
query33	452	223	193	193
query34	979	824	468	468
query35	402	398	346	346
query36	572	547	542	542
query37	117	82	70	70
query38	996	841	833	833
query39	477	512	482	482
query39_1	449	482	462	462
query40	198	92	80	80
query41	59	58	54	54
query42	75	73	76	73
query43	243	248	209	209
query44	999	530	544	530
query45	112	112	108	108
query46	775	857	535	535
query47	755	756	716	716
query48	319	323	220	220
query49	549	239	192	192
query50	722	268	203	203
query51	8294	8252	8082	8082
query52	67	69	62	62
query53	189	208	144	144
query54	218	167	156	156
query55	67	62	54	54
query56	218	177	147	147
query57	794	657	673	657
query58	189	178	181	178
query59	1200	1223	1116	1116
query60	242	197	177	177
query61	128	129	132	129
query62	367	203	176	176
query63	179	152	145	145
query64	2769	742	717	717
query65	1581	1558	1590	1558
query66	1828	307	239	239
query67	9738	9735	9792	9735
query68	2966	1231	697	697
query69	341	266	189	189
query70	675	616	624	616
query71	249	178	161	161
query72	2254	1702	1465	1465
query73	613	575	338	338
query74	1981	1224	1151	1151
query75	1173	1104	954	954
query76	2360	732	523	523
query77	253	257	196	196
query78	4078	3760	3154	3154
query79	2762	805	563	563
query80	1545	318	274	274
query81	504	161	137	137
query82	626	121	100	100
query83	285	206	188	188
query84	307	106	86	86
query85	786	348	285	285
query86	398	178	180	178
query87	1031	1006	910	910
query88	2888	2068	2084	2068
query89	283	196	177	177
query90	1936	127	133	127
query91	131	119	99	99
query92	82	69	71	69
query93	1603	1056	676	676
query94	644	256	207	207
query95	528	312	237	237
query96	800	601	271	271
query97	1052	1083	998	998
query98	172	135	135	135
query99	413	349	307	307
Total cold run time: 178423 ms
Total hot run time: 81783 ms
```
</details>

**Comment by hello-stephen** [comment]
<details>
<summary>ClickBench: <b>Total hot run time: 14.83 s</b></summary>

```
machine: 'aliyun_ecs.c7a.8xlarge_32C64G'
scripts: https://github.com/apache/doris/tree/master/tools/clickbench-tools
ClickBench test result on commit d82e48a99dc60c4e692e53ed141210bcbf362df5, data reload: false

query1	0.00	0.01	0.00
query2	0.07	0.04	0.04
query3	0.24	0.11	0.11
query4	1.60	0.10	0.09
query5	0.18	0.16	0.16
query6	1.27	0.71	0.68
query7	0.03	0.01	0.00
query8	0.04	0.04	0.03
query9	0.28	0.21	0.21
query10	0.36	0.36	0.34
query11	0.16	0.11	0.12
query12	0.15	0.12	0.13
query13	0.30	0.31	0.32
query14	0.46	0.44	0.46
query15	0.36	0.36	0.34
query16	0.22	0.21	0.22
query17	0.73	0.73	0.73
query18	0.17	0.17	0.15
query19	1.21	1.21	1.27
query20	0.01	0.01	0.01
query21	15.45	0.17	0.11
query22	5.09	0.05	0.04
query23	16.16	0.26	0.10
query24	2.99	0.29	0.25
query25	0.10	0.04	0.03
query26	0.84	0.17	0.13
query27	0.03	0.03	0.03
query28	3.65	0.55	0.27
query29	12.46	3.21	2.62
query30	0.25	0.12	0.12
query31	2.76	0.38	0.17
query32	3.52	0.32	0.23
query33	1.44	1.48	1.42
query34	15.39	2.18	1.79
query35	1.79	1.75	1.72
query36	0.46	0.30	0.29
query37	0.06	0.04	0.03
query38	0.05	0.03	0.03
query39	0.03	0.02	0.03
query40	0.11	0.08	0.08
query41	0.08	0.02	0.02
query42	0.03	0.02	0.02
query43	0.04	0.03	0.02
Total cold run time: 90.62 s
Total hot run time: 14.83 s
```
</details>

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `100.00% (4/4)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/67896_d82e48a99dc60c4e692e53ed141210bcbf362df5_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/67896_d82e48a99dc60c4e692e53ed141210bcbf362df5_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 76.28% (34616/45380) |
| Line Coverage     | 61.26% (389293/635491) |
| Region Coverage   | 57.48% (326590/568198) |
| Branch Coverage   | 58.30% (148959/255526) |

**Review by Mryange — APPROVED** [review]
